### DISCUSSION TITLE
Why do I get different behavior in CSharpScript than my code?

### DISCUSSION THREAD

**Discussion by Nefcanto** [dotnet/runtime#110037] [Q&A]
I have a codebase (an ASP.NET Core API) that works. Now I want to run some C# scripts, based on this codebase.

This is my code to run the C# script:

```
            var usings = GetUsings(); // the same for the app, a list of "global using X;"
            var assemblies = AppDomain
                .CurrentDomain
                .GetAssemblies()
                .Where(asm => !asm.IsDynamic && !string.IsNullOrWhiteSpace(asm.Location))
                .ToList();
            var options = ScriptOptions
                .Default
                .AddImports(usings)
                .AddReferences(assemblies)
                .AddReferences(typeof(Microsoft.Data.SqlClient.SqlConnection).Assembly)
                .AddReferences(Assembly.Load("MoreLinq"))
                .AddReferences(Assembly.Load("Pomelo.EntityFrameworkCore.MySql.NetTopologySuite"))
                ;
            var result = CSharpScript.EvaluateAsync(code, options, globals: globals).Result;
```

Now I get this error:

> error CS0234: The type or namespace name 'Imaging' does not exist in the namespace 'System.Drawing' (are you missing an assembly reference?)

I'm stuck here. I'm on Linux (Ubuntu 24.04) and inside Docker (Debian 12). I'm using .NET 8.

The same code, the same set of assemblies, and `using` statements work when I run my API. And this code snippet runs inside that API.

Why do I get two different behaviors? How should I solve this issue? Why do I not get that error when I run my API?

**Comment by Nefcanto** [comment]
To put it another way, inside my code (which works) I try to add all of the using statements and all assemblies to a C# script evaluator to run a C# script. Why it breaks?

**Comment by huoyaoyuan** [comment]
It's a common mistake to get reference assemblies from `AppDomain.GetAssemblies()`. It contains only the assemblies has been loaded. You need to use either the SDK pack, or the whole runtime pack to get everything that's available.

**Reply by jaredpar** [reply]
It is also a mistake because the runtime assemblies are not designed to be compiled against, they are designed to be run against. Compiling code against runtime assemblies can, and absolutely will, break between releases of .NET.

**Comment by Nefcanto** [comment]
@huoyaoyuan, I tried to use `Directory.GetFiles` to load all DLLs inside my bin and use `Assemly.LoadFrom` to load them. But that did not solve the issue either.

How can I add the SDK pack? Or runtime pack. To be hosent, I don't even know what those terms mean.

**Reply by huoyaoyuan** [reply]
> all DLLs inside my bin

It doesn't contain any framework assemblies if you are not doing standalone deployment.

SDK pack are assemblies under `dotnet\packs\Microsoft.NETCore.App.Ref\X.0.0\ref\netX.0`, they are the assemblies referenced by SDK to compile an application.
Runtime pack are assemblies under `dotnet\shared\Microsoft.NETCore.App\X.0.0`, they are loaded for execution. In standalone deployment, they will be copied together with the application.

Another approach is to use `typeof(SomeType).Assembly` to proactively load the types/assemblies your script may use.

**Reply by Nefcanto** [reply]
@huoyaoyuan, is there an easy way to tell the C# script evaluator to use the same context as this code that is being run? Something like `ScriptOptions.Default.TheSameAsTheCaller()`?

**Reply by huoyaoyuan** [reply]
`AppDomain.GetAssemblies()` is basically the same context *at that moment*. However in .NET, loading is lazy, so everything not used until present time won't have been loaded. That's why proactively touch the assemblies via *types in them* is required.

**Reply by Nefcanto** [reply]
@huoyaoyuan, this is the point. All I want as a developer is to be able to run some C# code dynamically. To achieve this, I need to understand a lot about the assemblies, and the underlying way .NET works. You are good at that. But .NET consumers are not good at it. We are good at tackling business requirements, using the technologies you have provided.

So, I think it should be a logical request to ask the .NET team to make us run a code, just as it was running statically inside the codebase, against the current app, and make it easy for us. A flag should be there that I can set to true, and then you do all the heavy work of finding what and where, etc.

Where can I ask for this feature? Because I'm stuck and kindly, I can't use the information you provided to solve my issue.

**Reply by huoyaoyuan** [reply]
I'm aware of the requirement of knowledge when I was looking for the solution. I'm not sure whom is to ping for this. Maybe @jaredpar to help direct this to the right person?

> A flag should be there that I can set to true, and then you do all the heavy work of finding what and where, etc.

Be aware of the nature of the feature request. It will make you incompatible with features like trimming.

> the same context as this code that is being run

Highlight this again: you can't really be on the same context, because the code itself is impacting the context during building and deploying. All the approaches can only take you close to the context. This is the nature of non-scripting language.

**Reply by jaredpar** [reply]
> A flag should be there that I can set to true, and then you do all the heavy work of finding what and where, etc.

Part of the reason this doesn't exist is that there isn't always an easy answer. Build is _complicated_ and in order to compile + eval your application has to take on this complexity. For some subset of scenarios like compile against only assemblies in a given target framework there are fairly straight forward answers. The moment you start adding other assemblies to the equation though it gets complicated for all the reasons build is complicated.

In terms of having a simple eval loop against standard reference assemblies, you can use the code in [roslyn-compile-and-run](https://github.com/jaredpar/roslyn-compile-and-run) as a template to build off of. But your example uses other assemblies, including custom ones, integrating them might be easy or might be more challenging depending on their contents. 

If you want to post the specific assemblies and the setup we can likely give better guidance.

**Reply by Nefcanto** [reply]
@jaredpar, when you say "complicated" which I assume you mean "complex" I understand. Anybody who is fair enough knows that software is complex by nature.
Yet my quest is very specific. The current code goes through static completion against the SDK, and the final result works against the runtime. Let's say you have the option to consider the boolean I have passed in, and if that boolean exists and is true, then you can copy **ALL** of the assemblies that you deem necessary to run the current app inside a directory. Let's call that directory **ScriptDependencies**. Why can you do that? Because you're compiling our code already. You have already done that **complex** task. You have analyzed the code. You have gone the path.

You know *what is necessary to run this app*. Just copy that into a directory and load it again when the script runs.

Does that work?

**Reply by jaredpar** [reply]
> when you say "complicated" which I assume you mean "complex" I understand.

No I meant complicated. 

> You know what is necessary to run this app

No. We know what is necessary to _compile_ the application. There is a difference between _compile_ (the act of turning source files into diagnostics and / or DLLs) and _deploy_ (the act of taking the output of compile and making it runnable). 

> Why can you do that? 

This is also something that you could do in a post-compile step. The set of `@(References)` is available for any target that you run after compile and it could be used to copy them to a directory in your output folder.

**Reply by Nefcanto** [reply]
@jaredpar, I may not have learned those terms correctly. No offense, but the way I learned them, *complex* is by nature, *complicated* is that we made a mess. Complex is what it is, complicated is what we made it to be. Have I learned them right? If so, then why it's complicated?

Also, again let me know what I lack here. When you compile an app, you find all of the dependencies and the target platform to do so. You know that:

> This guy needs Morelinq.dll because he has used `Where(filterString)` and he's going to run his app on Linux. Therefore I should pick this specific Morelinq.dll and bundle it with his application.

The last point seems to be interesting. Could you please provide a link or explain it more?

**Reply by spydacarnage** [reply]
> I may not have learned those terms correctly. No offense, but the way I learned them, complex is by nature, complicated is that we made a mess. Complex is what it is, complicated is what we made it to be. Have I learned them right? If so, then why it's complicated?

The easiest differentiation of complex vs. complicated is that something is complicated when it becomes difficult to understand in itself, but it becomes complex when it made of many, possibly simple, things that are difficult to understand due to the nature of their interactions.

**Reply by jaredpar** [reply]
>  When you compile an app, you find all of the dependencies and the target platform to do so.

During compile you find all of the dependencies necessary to _build_ the application. That is not necessarily the set of dependencies necessary to execute the application. .NET makes a distinction between reference and implementation. 

> No offense, but the way I learned them, complex is by nature, complicated is that we made a mess

Can apply either to build and it works just fine.

**Reply by Nefcanto** [reply]
@jaredpar, so what do you propose I do? I want to execute some C# script as if it were part of the currently compiled application. I still don't get it why when I compile and run my app which includes `global using System.Drawing.Imaging;` it woks just fine, but when I add the same using statement to the script evaluator, it throws that error.

I'm stuck here. Please guide me.

The reason I want everything from the current static codebase that's compiled is because I have tons of infra code in it. Database utilities, validations, a lot of helpers, extension methods, etc. I don't want to repeat all of them in my C# script. I want to load the script from the database, and execute them just as if they were written statically in my code.

**Reply by davidfowl** [reply]
You can use [Microsoft.Extensions.DependencyModel](https://www.nuget.org/packages/Microsoft.Extensions.DependencyModel#readme-body-tab) to accomplish this. This is what razor runtime compilation does. There are instructions on the readme for how to use it.

This will preserve all of the assemblies used to compile your project at runtime, then you can use these APIs to get access to that information.

**Reply by Nefcanto** [reply]
@davidfowl, now that's something valuable. Thank you so much for proposing it. I'll give it a try. But I still wish there could be a simple flag. `SameAsCompiledApp = true` and the heavy lifting would be done by the .NET team.

**Reply by davidfowl** [reply]
Turns out compiling code at runtime is more complicated than it seems on the surface. Reference assemblies are the tip of the iceberg
